tensorflow/java/src/test/java/org/tensorflow/ScopeTest.java

```diff
@@ -0,0 +1,184 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+package org.tensorflow;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Unit tests for {@link org.tensorflow.Scope}. */
+@RunWith(JUnit4.class)
+public class ScopeTest {
+
+  @Test
+  public void basic() {
+    try (Graph g = new Graph()) {
+      Scope s = Scope.create(g);
+      Const c1 = Const.create(s, 42);
+      assertEquals("Const", c1.input().op().name());
+      Const c2 = Const.create(s, 7);
+      assertEquals("Const_1", c2.input().op().name());
+      Const c3 = Const.create(s.withOpName("four"), 4);
+      assertEquals("four", c3.input().op().name());
+    }
+  }
+
+  @Test
+  public void hierarchy() {
+    try (Graph g = new Graph()) {
+      Scope root = Scope.create(g);
+      Scope child = root.withSubScope("child");
+      assertEquals("child/Const", Const.create(child, 42).input().op().name());
+      assertEquals("child/four", Const.create(child.withOpName("four"), 4).input().op().name());
+    }
+  }
+
+  @Test
+  public void composite() {
+    try (Graph g = new Graph();
+        Session sess = new Session(g)) {
+      Scope s = Scope.create(g);
+      Const data = Const.create(s.withOpName("data"), new int[] {600, 470, 170, 430, 300});
+
+      // Create a composite op with a customized name
+      Variance var1 = Variance.create(s.withOpName("example"), data);
+      assertEquals("example/variance", var1.input().op().name());
+
+      // Confirm internally added ops have the right names.
+      assertNotNull(g.operation("example/squared_deviation"));
+      assertNotNull(g.operation("example/Mean"));
+      assertNotNull(g.operation("example/zero"));
+
+      // Same composite op with a default name
+      Variance var2 = Variance.create(s, data);
+      assertEquals("variance/variance", var2.input().op().name());
+
+      // Confirm internally added ops have the right names.
+      assertNotNull(g.operation("variance/squared_deviation"));
+      assertNotNull(g.operation("variance/Mean"));
+      assertNotNull(g.operation("variance/zero"));
+
+      // Verify correct results as well.
+      Tensor result = sess.runner().fetch(var1.input()).run().get(0);
+      assertEquals(21704, result.intValue());
+      result = sess.runner().fetch(var2.input()).run().get(0);
+      assertEquals(21704, result.intValue());
+    }
+  }
+
+  // Convenience interface
+  // TODO: replace when standardized
+  interface Input {
```

Comment:
For now, I know we might undo this, but might I recommend getting rid of this interface and the `input()` method in the "handwritten" classes? Without the complete context, things like `public Output input() { return output; }` and the use in `Const(...).input().op()...` seem confusing. Better to keep this patch focused on the classes it adds?